### ISSUE TITLE
Critical Bug: Artificial  Equity Floor Prevents Ruin Detection in Monte Carlo Simulations

### ISSUE THREAD

**Issue by AlexFiliakov** [AlexFiliakov/Ergodic-Insurance-Limits#207]
# 🐛 Critical Bug: Artificial $1 Equity Floor Prevents Ruin Detection in Monte Carlo Simulations

## Description

Three interconnected bugs artificially prevent companies from becoming insolvent during Monte Carlo simulations, invalidating all ruin probability calculations and ergodic growth rate analysis:

1. **$1 Equity Floor in Loss Absorption** - Companies cannot absorb losses below $1 equity
2. **Incorrect Ruin Detection Threshold** - Checks for `<= 0` but equity never reaches 0
3. **Float32 Precision Loss** - Compounds numerical issues

**Impact**: All simulations show 0% ruin rate regardless of risk level. Research results showing optimal insurance strategies are invalid.

## Context

- **Project Area**: Core simulation engine (Monte Carlo, Manufacturer model)
- **Complexity**: 5/10 (Multiple files, simplified to single tolerance parameter)
- **Affects**: manufacturer.py, monte_carlo.py, simulation.py + 6 test files
- **Priority**: 🔴 **Critical** - Blocks research validation

## Current Behavior

```python
# Example from tail simulation results
final_assets = np.float32(1.0)  # Stops at exactly $1
ruin_probability = 0.0  # Always zero despite extreme risk scenarios
```

Companies with billions in losses and $1 remaining equity are marked as "survived."

## Expected Behavior

- Companies should reach equity ≈ $0 when insolvent
- Ruin detection should trigger when equity <= $10,000 (configurable)
- Simulations should stop when insolvency is detected
- Ruin probabilities should reflect actual insolvency rates

### Bug #1: Loss Absorption Buffer (manufacturer.py:1379-1384)

```python
# CURRENT CODE (INCORRECT):
max_loss = (
    min(abs(retained_earnings), current_equity - 1.0, available_cash)
    if (current_equity > 1.0 and available_cash > 0)
    else 0.0
)
```

**Problem**: Hardcoded 1.0 buffer creates artificial floor.

**Fix**: Use single configurable tolerance:
```python
tolerance = self.config.insolvency_tolerance  # default 10_000
max_loss = (
    min(abs(retained_earnings), current_equity - tolerance, available_cash)
    if (current_equity > tolerance and available_cash > 0)
    else 0.0
)
```

### Bug #2: Ruin Detection Threshold (monte_carlo.py:143, 1005)

```python
# CURRENT CODE (INCORRECT):
if manufacturer.total_assets <= 0:
    # Mark ruin and break
```

**Problem**: Assets never reach 0 due to Bug #1.

**Fix**: Check against tolerance (same value):
```python
tolerance = self.config.insolvency_tolerance  # default 10_000
if manufacturer.equity <= tolerance:
    is_ruined = True
    break  # Stop simulation
```

### Bug #3: Float32 Default (monte_carlo.py:186)

```python
# CURRENT CODE (SUBOPTIMAL):
use_float32: bool = True
```

**Fix**: Change default to False for critical calculations:
```python
use_float32: bool = False  # Use float64 for accuracy
```

## Requirements

1. **Add Single Configuration Parameter**:
   - `insolvency_tolerance` (default: 10_000) - Single threshold for both loss absorption and ruin detection
   - Add to ManufacturerConfig and SimulationConfig
   - **Rationale**: A company with $10k equity (0.1% of $10M starting assets) is effectively insolvent

2. **Update Loss Absorption Logic**:
   - Replace hardcoded 1.0 buffer with `insolvency_tolerance`
   - Update manufacturer.py lines 1365-1384

3. **Update Ruin Detection**:
   - Use same `insolvency_tolerance` threshold
   - Check `equity <= insolvency_tolerance` (simpler than checking both assets and equity)
   - Update monte_carlo.py lines 143, 1005
   - Update simulation.py lines 711, 806

4. **Change Float Precision Default**:
   - Set `use_float32 = False` in SimulationConfig

5. **Add Validation**:
   - Assert equity reaches near-zero when appropriate
   - Log warnings when insolvency is detected
   - Validate ruin_probability > 0 in high-risk scenarios

## Implementation Steps

- [ ] 1. Add `insolvency_tolerance: float = 10_000` to `config_v2.py` ManufacturerConfig
- [ ] 2. Add `insolvency_tolerance: float = 10_000` to `monte_carlo.py` SimulationConfig
- [ ] 3. Update loss absorption in `manufacturer.py:1379-1384` to use `self.config.insolvency_tolerance` instead of hardcoded 1.0
- [ ] 4. Update equity floor check in `manufacturer.py:1365` to use `insolvency_tolerance`
- [ ] 5. Update ruin detection in `monte_carlo.py:143` to check `equity <= insolvency_tolerance`
- [ ] 6. Update ruin detection in `monte_carlo.py:1005` to check `equity <= insolvency_tolerance`
- [ ] 7. Update ruin detection in `simulation.py:711` to check `equity <= insolvency_tolerance`
- [ ] 8. Update ruin detection in `simulation.py:806` to check `equity <= insolvency_tolerance`
- [ ] 9. Change `use_float32` default to `False` in `monte_carlo.py:186`
- [ ] 10. Add validation logic in `check_solvency()` to log when equity ≈ insolvency_tolerance
- [ ] 11. Update `test_ruin_probability.py` with new tolerance parameter
- [ ] 12. Update `test_manufacturer.py` to test insolvency_tolerance behavior
- [ ] 13. Update `test_manufacturer_methods.py` for loss absorption tests
- [ ] 14. Update `test_monte_carlo.py` for ruin detection tests
- [ ] 15. Update `test_periodic_ruin_tracking.py` for tolerance-based detection
- [ ] 16. Update `test_simulation.py` for insolvency tests
- [ ] 17. Add integration test validating ruin occurs in high-risk scenarios
- [ ] 18. Update documentation explaining insolvency_tolerance parameter
- [ ] 19. Run full test suite and fix any failures
- [ ] 20. Validate existing notebooks produce non-zero ruin probabilities

## Acceptance Criteria

- [ ] Single `insolvency_tolerance` parameter used throughout codebase
- [ ] Default value of $10,000 represents practical insolvency threshold
- [ ] Companies can reach equity values between $0 and $10,000
- [ ] Ruin detection triggers when equity <= insolvency_tolerance
- [ ] Simulations stop when insolvency is detected
- [ ] High-risk simulations show non-zero ruin probabilities
- [ ] All existing tests pass with updated logic
- [ ] New tests validate ruin occurs in extreme loss scenarios
- [ ] Configuration parameter is properly documented
- [ ] Backwards compatibility: old code works with default tolerance

## Testing Requirements

```python
# Test 1: Verify ruin occurs with extreme losses
def test_ruin_detection_with_extreme_losses():
    config = ManufacturerConfig(
        initial_assets=1_000_000,
        insolvency_tolerance=10_000
    )
    manufacturer = WidgetManufacturer(config)
    # Apply $1.5M loss
    manufacturer.process_uninsured_claim(1_500_000)
    assert manufacturer.is_ruined == True
    assert manufacturer.equity <= config.insolvency_tolerance

# Test 2: Verify equity can reach near-zero
def test_equity_reaches_near_insolvency_threshold():
    config = ManufacturerConfig(
        initial_assets=100_000,
        insolvency_tolerance=10_000
    )
    manufacturer = WidgetManufacturer(config)
    # Apply $95k loss - should leave ~$5k equity
    manufacturer.process_uninsured_claim(95_000)
    assert 0 < manufacturer.equity < config.insolvency_tolerance
    assert manufacturer.is_ruined == True  # Below threshold

# Test 3: Verify non-zero ruin in Monte Carlo
def test_monte_carlo_shows_ruin():
    config = SimulationConfig(
        n_simulations=10_000,
        insolvency_tolerance=10_000
    )
    # Run high-risk scenarios
    results = engine.run()
    assert results.ruin_probability[str(config.n_years)] > 0.01

# Test 4: Verify simulation stops at insolvency
def test_simulation_stops_at_insolvency():
    # Company should stop operating once equity <= insolvency_tolerance
    # Remaining years should be filled with zeros
    pass
```

## Files Affected

**Core Implementation** (4 files):
- `ergodic_insurance/config_v2.py` (add insolvency_tolerance parameter)
- `ergodic_insurance/manufacturer.py` (lines 1365, 1379-1384, 2462-2504)
- `ergodic_insurance/monte_carlo.py` (lines 143, 186, 1005)
- `ergodic_insurance/simulation.py` (lines 711, 806)

**Tests** (6+ files):
- `ergodic_insurance/tests/test_ruin_probability.py`
- `ergodic_insurance/tests/test_manufacturer.py`
- `ergodic_insurance/tests/test_manufacturer_methods.py`
- `ergodic_insurance/tests/test_monte_carlo.py`
- `ergodic_insurance/tests/test_periodic_ruin_tracking.py`
- `ergodic_insurance/tests/test_simulation.py`

**Documentation**:
- Config parameter docs
- Notebooks may show different results (document breaking change)

## Design Rationale: Why $10,000?

**Context**: Companies typically start with $10M in assets.

**$10,000 represents 0.1% of initial assets** - effectively insolvent:
- Cannot cover typical claim deductibles ($100k+)
- Cannot maintain operations (payroll, inventory, etc.)
- Cannot post collateral for insurance
- Practically bankrupt even if technically solvent

**Users can adjust** for different scenarios:
- Smaller companies: Lower tolerance (e.g., $1,000)
- Larger companies: Higher tolerance (e.g., $100,000)
- Research studies: Test sensitivity to threshold

## Breaking Changes

⚠️ **This fix will change existing simulation results**:
- Ruin probabilities will increase significantly (become more accurate)
- Companies will be marked as ruined earlier (at $10k equity instead of $0)
- Optimal insurance strategies may shift
- Existing research results need revalidation
- Notebooks will show different outputs

Document as breaking change in CHANGELOG and research papers.

## Related Discussion

See investigation thread that identified this bug: The $1 floor was originally intended to prevent floating-point rounding errors but inadvertently created a survival floor that invalidates all ruin calculations.
